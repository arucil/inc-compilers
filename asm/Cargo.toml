[package]
name = "asm"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
indexmap = "1.8.1"
enum-iterator = "0.8.1"
num-derive = "0.3.3"
num-traits = "0.2.15"
support = { path = "../support" }
