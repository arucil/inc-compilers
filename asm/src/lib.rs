#![feature(never_type)]

use num_derive::{FromPrimitive, ToPrimitive};
use std::fmt::{self, Debug, Formatter, Write};

#[derive(Debug, Clone)]
pub struct Program<INFO = (), VAR = !> {
  pub info: INFO,
  pub blocks: Vec<(String, Block<VAR>)>,
}

#[derive(Clone)]
pub struct Block<VAR = !> {
  pub code: Vec<Instr<VAR>>,
}

#[derive(Clone)]
#[non_exhaustive]
pub enum Instr<VAR = !> {
  Add(Arg<VAR>, Arg<VAR>),
  Sub(Arg<VAR>, Arg<VAR>),
  Mov(Arg<VAR>, Arg<VAR>),
  Neg(Arg<VAR>),
  Call(String, usize),
  Ret,
  Push(Arg<VAR>),
  Pop(Arg<VAR>),
  Jmp(String),
  Syscall,
}

#[derive(Clone, PartialEq, Eq)]
pub enum Arg<VAR = !> {
  Imm(i64),
  Reg(Reg),
  Deref(Reg, i32),
  Var(VAR),
}

#[derive(
  Clone, Copy, FromPrimitive, ToPrimitive, PartialEq, Eq, PartialOrd, Ord, Hash,
)]
pub enum Reg {
  Rsp,
  Rbp,
  Rax,
  Rbx,
  Rcx,
  Rdx,
  Rsi,
  Rdi,
  R8,
  R9,
  R10,
  R11,
  R12,
  R13,
  R14,
  R15,
}

impl<INFO: Debug, VAR: Debug> Program<INFO, VAR> {
  pub fn to_string_pretty(&self) -> String {
    let mut buf = format!("{:?}", self.info);
    for (label, block) in &self.blocks {
      writeln!(&mut buf, "{}:", label).unwrap();
      buf += &format!("{:?}", block);
    }
    buf
  }

  pub fn to_nasm(&self) -> String {
    let mut buf = format!(
      "extern read_int, print_int, print_newline\n\
      section .text\n"
    );
    for (label, block) in &self.blocks {
      buf += "\n";
      if label == "_start" {
        buf += "    global _start\n";
      }
      writeln!(&mut buf, "{}:", label).unwrap();
      buf += &format!("{:?}", block);
    }
    buf
  }
}

impl<VAR: Debug> Debug for Block<VAR> {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    for instr in &self.code {
      write!(f, "    ")?;
      instr.fmt(f)?;
      writeln!(f)?;
    }
    Ok(())
  }
}

impl<VAR: Debug> Debug for Instr<VAR> {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    match self {
      Self::Add(src, dest) => write!(f, "add {:?}, {:?}", dest, src),
      Self::Mov(src, dest) => write!(f, "mov {:?}, {:?}", dest, src),
      Self::Call(label, _) => write!(f, "call {}", label),
      Self::Jmp(label) => write!(f, "jmp {}", label),
      Self::Neg(dest) => write!(f, "neg {:?}", dest),
      Self::Pop(dest) => write!(f, "pop {:?}", dest),
      Self::Push(src) => write!(f, "push {:?}", src),
      Self::Ret => write!(f, "ret"),
      Self::Sub(src, dest) => write!(f, "sub {:?}, {:?}", dest, src),
      Self::Syscall => write!(f, "syscall"),
    }
  }
}

impl<VAR: Debug> Debug for Arg<VAR> {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    match self {
      Self::Imm(n) => write!(f, "{}", n),
      Self::Var(var) => var.fmt(f),
      Self::Deref(r, i) => {
        if *i > 0 {
          write!(f, "qword [{:?} + {}]", r, i)
        } else if *i == 0 {
          write!(f, "qword [{:?}]", r)
        } else {
          write!(f, "qword [{:?} - {}]", r, -i)
        }
      }
      Self::Reg(r) => r.fmt(f),
    }
  }
}

impl Reg {
  pub fn is_callee_saved(&self) -> bool {
    match self {
      Reg::Rsp
      | Reg::Rbp
      | Reg::Rbx
      | Reg::R12
      | Reg::R13
      | Reg::R14
      | Reg::R15 => true,
      _ => false,
    }
  }
}

impl Debug for Reg {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    use Reg::*;
    let op = match self {
      Rsp => "rsp",
      Rbp => "rbp",
      Rax => "rax",
      Rbx => "rbx",
      Rcx => "rcx",
      Rdx => "rdx",
      Rsi => "rsi",
      Rdi => "rdi",
      R8 => "r8",
      R9 => "r9",
      R10 => "r10",
      R11 => "r11",
      R12 => "r12",
      R13 => "r13",
      R14 => "r14",
      R15 => "r15",
    };
    f.write_str(op)
  }
}
