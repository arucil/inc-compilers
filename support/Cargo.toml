[package]
name = "support"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
assert_cmd = "2.0.4"
predicates = "2.1.1"
pretty_assertions = "1.2.1"
scopeguard = "1.1.0"
