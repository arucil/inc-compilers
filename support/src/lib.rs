pub mod compile_error;
pub mod range;
pub mod test_cli;

pub use self::compile_error::*;
pub use self::range::*;
pub use self::test_cli::*;
