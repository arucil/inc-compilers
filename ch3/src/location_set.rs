use asm::{Arg, Reg};
use ast::IdxVar;
use indexmap::map::Entry;
use indexmap::IndexMap;
use num_traits::{FromPrimitive, ToPrimitive};
use smallvec::SmallVec;
use std::fmt::{self, Debug, Write};
use std::ops::Deref;

#[derive(Debug, Clone)]
pub struct LocationSet(SmallVec<[u32; 2]>);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Location(usize);

#[derive(Debug, Clone)]
pub struct VarStore(IndexMap<IdxVar, Var>);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Var(usize);

impl LocationSet {
  pub fn new() -> Self {
    Self(SmallVec::new())
  }

  pub fn add_reg(&mut self, reg: Reg) {
    let i = reg_index(reg);
    if self.0.is_empty() {
      self.0.push(0);
    }
    self.0[i / 32] |= 1 << (i % 32);
  }

  pub fn remove_reg(&mut self, reg: Reg) {
    let i = reg_index(reg);
    if !self.0.is_empty() {
      self.0[i / 32] &= !(1 << (i % 32));
    }
  }

  pub fn add_var(&mut self, var: Var) {
    let i = var_index(var);
    let offset = i / 32;
    if offset >= self.0.len() {
      self.0.resize(offset + 1, 0);
    }
    self.0[offset] |= 1 << (i % 32);
  }

  pub fn remove_var(&mut self, var: Var) {
    let i = var_index(var);
    let offset = i / 32;
    if offset < self.0.len() {
      self.0[offset] &= !(1 << (i % 32));
    }
  }

  pub fn remove_caller_saved_regs(&mut self) {
    self.remove_reg(Reg::Rax);
    self.remove_reg(Reg::Rcx);
    self.remove_reg(Reg::Rdx);
    self.remove_reg(Reg::Rsi);
    self.remove_reg(Reg::Rdi);
    self.remove_reg(Reg::R8);
    self.remove_reg(Reg::R9);
    self.remove_reg(Reg::R10);
    self.remove_reg(Reg::R11);
  }

  pub fn add_argument_regs(&mut self, arity: usize) {
    const ARGUMENT_REGS: [Reg; 6] =
      [Reg::Rdi, Reg::Rsi, Reg::Rdx, Reg::Rcx, Reg::R8, Reg::R9];

    assert!(arity <= 6);
    for reg in &ARGUMENT_REGS[..arity] {
      self.add_reg(*reg);
    }
  }

  pub fn iter(&self) -> LocationIter {
    let current = if self.0.len() > 0 { self.0[0] } else { 0 };
    LocationIter {
      set: self,
      index: 0,
      current,
    }
  }

  pub fn write(&self, f: &mut impl Write, var_store: &VarStore) -> fmt::Result {
    write!(f, "{{")?;
    let mut comma = false;
    for elem in self {
      if comma {
        write!(f, ", ")?;
      }
      comma = true;
      write!(f, "{:?}", elem.to_arg(var_store))?;
    }
    write!(f, "}}")
  }
}

pub struct LocationIter<'a> {
  set: &'a LocationSet,
  index: usize,
  current: u32,
}

impl<'a> Iterator for LocationIter<'a> {
  type Item = Location;

  fn next(&mut self) -> Option<Location> {
    while self.current == 0 {
      self.index += 1;
      if self.index >= self.set.0.len() {
        return None;
      }
      self.current = self.set.0[self.index];
    }
    let i = self.current.trailing_zeros() as usize + self.index * 32;
    self.current &= self.current - 1;
    Some(Location(i))
  }
}

impl<'a> IntoIterator for &'a LocationSet {
  type Item = Location;
  type IntoIter = LocationIter<'a>;

  fn into_iter(self) -> Self::IntoIter {
    self.iter()
  }
}

fn reg_index(reg: Reg) -> usize {
  reg.to_usize().unwrap()
}

fn var_index(var: Var) -> usize {
  var.0 + 16
}

impl VarStore {
  pub fn new() -> Self {
    Self(IndexMap::new())
  }

  pub fn get(&mut self, var: IdxVar) -> Var {
    let len = self.0.len();
    match self.0.entry(var) {
      Entry::Occupied(var) => *var.get(),
      Entry::Vacant(entry) => *entry.insert(Var(len)),
    }
  }
}

impl Deref for VarStore {
  type Target = IndexMap<IdxVar, Var>;

  fn deref(&self) -> &Self::Target {
    &self.0
  }
}

impl Var {
  pub fn to_arg(&self, var_store: &VarStore) -> Arg<IdxVar> {
    Arg::Var(var_store.get_index(self.0 - 16).unwrap().0.clone())
  }
}

impl Location {
  pub fn to_arg(&self, var_store: &VarStore) -> Arg<IdxVar> {
    if self.0 < 16 {
      Arg::Reg(Reg::from_usize(self.0).unwrap())
    } else {
      Arg::Var(var_store.0.get_index(self.0 - 16).unwrap().0.clone())
    }
  }

  pub fn from_arg(arg: Arg<IdxVar>, var_store: &mut VarStore) -> Option<Self> {
    match arg {
      Arg::Deref(reg, _) | Arg::Reg(reg) => Some(reg.into()),
      Arg::Var(var) => Some(Self(var_store.get(var).0 + 16)),
      Arg::Imm(_) => None,
      Arg::ByteReg(_) => unimplemented!(),
    }
  }

  pub fn to_reg(&self) -> Option<Reg> {
    Reg::from_usize(self.0)
  }

  pub fn to_var(&self) -> Option<Var> {
    if self.0 >= 16 {
      Some(Var(self.0 - 16))
    } else {
      None
    }
  }
}

impl From<Reg> for Location {
  fn from(reg: Reg) -> Self {
    Self(reg.to_usize().unwrap())
  }
}
