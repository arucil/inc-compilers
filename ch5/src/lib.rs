#![feature(box_syntax)]

use asm::{Arg, Block, Instr, Label, Program, Reg};
use ch3::location_set::LocationSet;
use maplit::hashmap;
use support::CompileError;

pub mod pass;

pub fn compile(
  input: &str,
  regs: Option<&[Reg]>,
) -> Result<String, CompileError> {
  use Reg::*;
  let prog = ast::parse(input)?;
  let prog = self::pass::typecheck::typecheck(prog)?;
  let prog = ch4::pass::shrink::shrink(prog);
  let prog = ch4::pass::uniquify::uniquify(prog);
  let prog = self::pass::uncover_get::uncover_get(prog);
  let prog = ch2::pass::remove_complex_operands::remove_complex_operands(prog);
  let prog = ch4::pass::explicate_control::explicate_control(prog);
  let prog = ch4::pass::instruction_selection::select_instruction(prog, false);
  let prog = self::pass::liveness_analysis::analyze_liveness(
    prog,
    hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rbp])
    },
  );
  let prog = ch3::pass::interference::build_interference(prog);
  let prog = ch3::pass::move_biasing::build_move_graph(prog);
  let regs = regs.unwrap_or(&[
    Rbx, Rcx, Rdx, Rsi, Rdi, R8, R9, R10, R11, R12, R13, R14, R15,
  ]);
  let prog = ch3::pass::register_allocation::allocate_registers(prog, regs);
  let mut prog = ch2::pass::patch_instructions::patch_instructions(prog);
  add_prologue(&mut prog);
  add_epilogue(&mut prog);
  let prog = ch4::pass::merge_blocks::merge_blocks(prog);

  Ok(prog.to_nasm())
}

fn add_prologue(prog: &mut Program<ch3::pass::register_allocation::Info>) {
  use asm::Reg::*;
  use Instr::*;
  let stack_space =
    prog.info.stack_space + prog.info.used_callee_saved_regs.len() * 8;
  let stack_space =
    ((stack_space + 15) & !15) - prog.info.used_callee_saved_regs.len() * 8;
  let mut code = vec![
    Push(Arg::Reg(Rbp)),
    Mov {
      src: Arg::Reg(Rsp),
      dest: Arg::Reg(Rbp),
    },
    Sub {
      src: Arg::Imm(stack_space as i64),
      dest: Arg::Reg(Rsp),
    },
  ];
  for &reg in &prog.info.used_callee_saved_regs {
    code.push(Push(Arg::Reg(reg)));
  }
  code.push(Jmp(Label::Start));
  let block = Block { global: true, code };
  prog.blocks.push((Label::EntryPoint, block));
}

fn add_epilogue(prog: &mut Program<ch3::pass::register_allocation::Info>) {
  use asm::Reg::*;
  use Instr::*;
  let mut code: Vec<Instr> = prog
    .info
    .used_callee_saved_regs
    .iter()
    .rev()
    .map(|&reg| Pop(Arg::Reg(reg)))
    .collect();
  code.extend_from_slice(&[
    Mov {
      src: Arg::Reg(Rbp),
      dest: Arg::Reg(Rsp),
    },
    Pop(Arg::Reg(Rbp)),
    Mov {
      src: Arg::Imm(60),
      dest: Arg::Reg(Rax),
    },
    Mov {
      src: Arg::Imm(0),
      dest: Arg::Reg(Rdi),
    },
    Syscall,
  ]);
  let block = Block {
    global: false,
    code,
  };
  prog.blocks.push((Label::Conclusion, block));
}

#[cfg(test)]
mod tests {
  use super::*;
  use insta::assert_snapshot;

  #[test]
  fn if_form() {
    let prog = compile(r#"(print (if (eq? (read) 1) 42 0))"#, None).unwrap();
    assert_snapshot!(prog);
  }

  #[test]
  fn nested_if() {
    let prog = compile(
      r#"
(let ([x (read)]
      [y (read)])
  (print
    (if (if (< x 1)
          (eq? x 0)
          (eq? x 2))
      (+ y 2)
      (+ y 10))))"#,
      None,
    )
    .unwrap();
    assert_snapshot!(prog);
  }

  #[test]
  fn if_in_init() {
    let prog = compile(
      r#"(let ([x (if (>= (read) 3) 10 77)]) (print (if (not (eq? x 10)) 41 2)))"#,
      None,
    )
    .unwrap();
    assert_snapshot!(prog);
  }

  #[test]
  fn while_loop() {
    let prog = compile(
      r#"
(let ([x 1] [y 2])
  (while (< x (- y))
    (set! x (+ x y))
    (set! y (if (> x 10) (- x 1) x)))
  (print (begin (set! x (+ x 1)) (if (eq? x y) "x" "Y")) "abc" (not #t) "def")
  (+ 77 (- (let ([k (read)]) (+ k 1)) 4))
  (void))
      "#,
      None,
    )
    .unwrap();
    assert_snapshot!(prog);
  }
}
