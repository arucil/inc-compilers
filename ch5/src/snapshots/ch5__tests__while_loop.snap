---
source: ch5/src/lib.rs
expression: prog

---
extern read_int, print_int, print_str, print_bool, print_newline
section .rodata
    const_1 db `x`
    const_2 db `Y`
    const_3 db `abc`
    const_4 db `def`
section .text

    global _start
_start:
    push rbp
    mov rbp, rsp
    sub rsp, 8
    push rbx
    mov rbx, 1
    mov rcx, 2

block3:
    mov rdx, rcx
    neg rdx
    cmp rbx, rdx
    jl block6
    add rbx, 1
    cmp rbx, rcx
    je block1
    mov rsi, const_2
    mov rdx, 1
    call print_str

block0:
    mov rsi, const_3
    mov rdx, 3
    call print_str
    mov rax, 1
    xor rax, 1
    call print_bool
    mov rsi, const_4
    mov rdx, 3
    call print_str
    call print_newline
    call read_int
    mov rbx, rax
    add rbx, 1
    sub rbx, 4
    mov rax, 17
    pop rbx
    call print_int
    call print_newline
    mov rsp, rbp
    pop rbp
    mov rax, 60
    mov rdi, 0
    syscall

block1:
    mov rsi, const_1
    mov rdx, 1
    call print_str
    jmp block0

block6:
    add rbx, rcx
    cmp rbx, 10
    jg block4
    mov rcx, rbx
    jmp block3

block4:
    mov rcx, rbx
    sub rcx, 1
    jmp block3

